File: LeetCode/Medium/0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func distributeCoins(root *TreeNode) int {
    ret := 0
    
    dfs := func(node *TreeNode) int { return 0 }
    dfs = func(node *TreeNode) int { 
        if node == nil { return 0 }
        
        node.Val += dfs(node.Left) + dfs(node.Right) - 1
        ret += abs(node.Val)
        return node.Val
    }
    
    dfs(root)
    
    return ret
}

func abs(a int) int {
    if a < 0 { return -a }
    return a
    
}

File: LeetCode/Medium/0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ret = 0;
    
    public int distributeCoins(TreeNode root) {
        dfs(root);
        
        return ret;
    }
    
    private int dfs(TreeNode node){
        if (node == null) return 0;
        
        node.val += dfs(node.left) + dfs(node.right);
        
        ret += Math.abs(node.val - 1);
        return node.val - 1;
    }
}